src/bio/smd/smd_device.c

```diff
@@ -230,112 +206,82 @@ fetch_dev_info(uuid_t dev_id, struct smd_dev_info **dev_info)
 int
 smd_dev_get_by_id(uuid_t dev_id, struct smd_dev_info **dev_info)
 {
-	int	rc;
+	struct d_uuid	id;
+	int		rc;
 
-	smd_lock(&smd_store);
-	rc = fetch_dev_info(dev_id, dev_info);
-	smd_unlock(&smd_store);
+	uuid_copy(id.uuid, dev_id);
+	smd_db_lock();
+	rc = smd_dev_get_info(&id, dev_info);
+	smd_db_unlock();
 	return rc;
 }
 
 int
-smd_dev_get_by_tgt(int tgt_id, struct smd_dev_info **dev_info)
+smd_dev_get_by_tgt(uint32_t tgt_id, struct smd_dev_info **dev_info)
 {
-	struct d_uuid	bond_dev;
-	d_iov_t		key, val;
+	struct d_uuid	id;
 	int		rc;
 
-	D_ASSERT(!daos_handle_is_inval(smd_store.ss_tgt_hdl));
-	smd_lock(&smd_store);
-
-	d_iov_set(&key, &tgt_id, sizeof(tgt_id));
-	d_iov_set(&val, &bond_dev, sizeof(bond_dev));
-	rc = dbtree_fetch(smd_store.ss_tgt_hdl, BTR_PROBE_EQ,
-			  DAOS_INTENT_DEFAULT, &key, NULL, &val);
+	smd_db_lock();
+	rc = smd_db_fetch(TABLE_TGT, &tgt_id, sizeof(tgt_id), &id, sizeof(id));
 	if (rc) {
 		D_CDEBUG(rc != -DER_NONEXIST, DLOG_ERR, DB_MGMT,
 			 "Fetch target %d failed. "DF_RC"\n", tgt_id,
 			 DP_RC(rc));
 		goto out;
 	}
-
-	rc = fetch_dev_info(bond_dev.uuid, dev_info);
+	rc = smd_dev_get_info(&id, dev_info);
 out:
-	smd_unlock(&smd_store);
+	smd_db_unlock();
 	return rc;
 }
 
-int
-smd_dev_list(d_list_t *dev_list, int *devs)
+static int
+smd_dev_list_cb(struct sys_db *db, char *table, d_iov_t *key, void *args)
 {
-	struct smd_dev_info	*info;
-	struct smd_dev_entry	 entry = { 0 };
-	daos_handle_t		 iter_hdl;
-	struct d_uuid		 key_dev;
-	d_iov_t			 key, val;
-	int			 dev_cnt = 0;
-	int			 rc;
-
-	D_ASSERT(dev_list && d_list_empty(dev_list));
-	D_ASSERT(!daos_handle_is_inval(smd_store.ss_dev_hdl));
-
-	smd_lock(&smd_store);
+	struct smd_trav_data    *td = args;
+	struct smd_dev_info     *info;
+	struct smd_device        dev;
+	struct d_uuid            id;
+	int                      rc;
+
+	D_ASSERT(key->iov_len == sizeof(id));
+	id = *(struct d_uuid *)key->iov_buf;
+	rc = smd_db_fetch(TABLE_DEV, &id, sizeof(id), &dev, sizeof(dev));
+	if (rc)
+		return rc;
 
-	rc = dbtree_iter_prepare(smd_store.ss_dev_hdl, 0, &iter_hdl);
-	if (rc) {
-		D_ERROR("Prepare device iterator failed. "DF_RC"\n", DP_RC(rc));
-		goto out;
-	}
+	info = smd_dev_alloc_info(&id, &dev);
+	if (!info)
+		return -DER_NOMEM;
 
-	rc = dbtree_iter_probe(iter_hdl, BTR_PROBE_FIRST, DAOS_INTENT_DEFAULT,
-			       NULL, NULL);
-	if (rc) {
-		if (rc != -DER_NONEXIST)
-			D_ERROR("Probe first device failed. "DF_RC"\n",
-				DP_RC(rc));
-		else
-			rc = 0;
-		goto done;
-	}
+	d_list_add_tail(&info->sdi_link, &td->td_list);
+	td->td_count++;
+	return 0;
+}
 
-	d_iov_set(&key, &key_dev, sizeof(key_dev));
-	d_iov_set(&val, &entry, sizeof(entry));
+int
+smd_dev_list(d_list_t *dev_list, int *devs)
```

Comment:
Locking is missed for listing function? Also, I'm wondering if we'd implicitly do locking for each individual operation in SYSDB? (since the single SYSDB instance is shared by all xstreams, locking is required for every tree operation). The user of SYSDB, like SMD, may introduce its own lock to ensure the atomicity of multiple SYSDB operations?

Reply:
Indeed the locking is missing, thanks.
I've put the lock in vos_sys_db is because I think we could use the same pmem pool for other purposes like exporting the TX log system to metadata operations, those user cases could be from different threads, so we have to create the lock which is associated with the sysdb, instead of putting it in SMD.